### ISSUE TITLE
Avoid duplicate information between the readme file and the bnd file

### ISSUE THREAD

**Issue by cdecode** [bndtools/bnd#2123]
README.md files contains a lot of information that are also present in the bundle manifest such as the bundle name, the description, the developer contacts, licences...

In order to keep this information in sync, the enroute project puts some variables in the readme file : 

> \# ${Bundle-Name}
${Bundle-Description}

This works and the jar contains a proper readme file when include with brackets, but the variables remain unchanged when we access the readme file from GitHub.

I do the reverse by using html tags and bnd macros to extract the manifest headers value, for exemple: 

> \# \<bnd-name\>Great Bundle\</bnd-name\> --> Bundle-Name: Great Bundle

This works but users must know the html tags that I used and the macros are a little complicated. 

A new bnd instruction such as **-readme: true**  could be more useful. This instruction could automatically create/update the readme file at the root of the project/workspace based on the manifest headers and include the readme in the jar at the same time.

Is this feature could be useful ?

**Comment by cdecode** [comment]
I have implemented the readme generation tool, an example of a readme file generated by the test case is as follow:

``` 
<bnd-name>

# My Bundle</bnd-name><bnd-description>

A super bundle</bnd-description>


<bnd-services>

# Components / Services

List of configurable services:

#### Pid: com.test *required*

|Name	|Type	|Description	|Default	|
|---	|---	|---	|---	|
|test.test	|String	|A test config	|truc	|
|test.test2	|String	|A test config	|""	|
|test.test2	|String	|A test config	|-	|


###### Properties

|Name	|Type	|Value	|
|---	|---	|---	|
|test.test	|String	|truc	|


###### Accessibility

|Interface	|
|---	|
|com.test.api.test	|
|com.test.api.test2	|


#### Pid: com.test2 *no configuration*

###### Properties

|Name	|Type	|Value	|
|---	|---	|---	|
|test.test	|String	|truc	|
</bnd-services>


<bnd-contacts>

# Contacts

Feel free to contact us: 

|Name	|Email	|Role	|
|---	|---	|---	|
|bob	|[bob@bob.com](mailto:bob@bob.com)	|President	|
|charles	|[charles@charles.com](mailto:charles@charles.com)	|Dev	|
</bnd-contacts>

<bnd-license>

# License

[Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)

Do what you want to do.</bnd-license>

<bnd-foot>

---
MyCompagny - 1.0.0</bnd-foot>
```

It look like this:

<bnd-name>

# My Bundle</bnd-name><bnd-description>

A super bundle</bnd-description>


<bnd-services>

# Components / Services

List of configurable services:

#### Pid: com.test *required*

|Name	|Type	|Description	|Default	|
|---	|---	|---	|---	|
|test.test	|String	|A test config	|truc	|
|test.test2	|String	|A test config	|""	|
|test.test2	|String	|A test config	|-	|


###### Properties

|Name	|Type	|Value	|
|---	|---	|---	|
|test.test	|String	|truc	|


###### Accessibility

|Interface	|
|---	|
|com.test.api.test	|
|com.test.api.test2	|


#### Pid: com.test2 *no configuration*

###### Properties

|Name	|Type	|Value	|
|---	|---	|---	|
|test.test	|String	|truc	|
</bnd-services>


<bnd-contacts>

# Contacts

Feel free to contact us: 

|Name	|Email	|Role	|
|---	|---	|---	|
|bob	|[bob@bob.com](mailto:bob@bob.com)	|President	|
|charles	|[charles@charles.com](mailto:charles@charles.com)	|Dev	|
</bnd-contacts>

<bnd-license>

# License

[Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)

Do what you want to do.</bnd-license>

<bnd-foot>

---
MyCompagny - 1.0.0</bnd-foot>

--End of file

and could be configured in the build.bnd like this:

``` 
-readme: \
    services-message: List of configurable services;\
    contacts-message: Feel free to contact us;\
    show-copyright: false   
```

What I need to know is how to get the source of information needed to generate the readme ([link to the method](https://github.com/cdecode/bnd/blob/18d9661f2d3811240a5aa492b4dcc56e26e38f29/biz.aQute.bndlib/src/aQute/bnd/readme/ReadmeInformation.java#L84)) and how to integrate an instruction in the bnd build process?!

**Comment by bjhargrave** [comment]
The Bnd devs discussed this today. We have some comments.

While the desire to generate a readme from the information in the bundle is a great idea. The scope of the proposed solution is too narrow. We would rather see a solution which generated metadata about the bundle as part of the build process. This could be placed in a resource added to the bundle under OSGI-OPT or it could even be a file written to the output folder. The information in this file is then useful to generate a readme or many other interesting things.

@pkriens suggested that the analysis of the bundle to produce this metadata could be done to allow it to be run on an already existing bundle. For example, the bnd command could be used to generate this from a jar file.

So basically, we see a 2 step process: (1) one collect metatdata about a bundle and write it out in some form (e.g. json, xml, properties file), (2) a separate step (like a Bnd plugin) to take that metadata as input and use it to generate a readme. The step (2) could be implemented in many ways include templating support for the readme generation, gradle or maven tasks, etc.

**Comment by cdecode** [comment]
Thanks to consider this feature, I have some questions regarding the metadata file:

* Which data should be represented ? The jar itself contains a lot of structured metadata (manifest, xml components...), does this file must contain all of this information such as package export...
* For the step (1), is the Builder class a good place for this ?
* I checked the VerifierPlugin for the step (2), can we include the readme in the jar at this step ? 

I can help for those tasks but I will need some guide lines on how bnd works (is there some docs for this?) and how you want to implement this.

very enthousiaste for this feature!

**Comment by cdecode** [comment]
I propose to create a new issue related to the metadata file, and when ready, create another pull request for this. We can get back on this issue for the readme generation then.

Is it fine ?

**Comment by bjhargrave** [comment]
That would be fine. As for the metadata. you should start with the information you collected for the readme generated and we can discuss from there. We should use an extensible and flexible format for the metadata storage. JSON may be appropriate. Bnd already has some JSON support in the aQute.lib.json package.

**Comment by cdecode** [comment]
If we use templating, maybe the bnd project could just include a templating plugin which can work on any file and the readme specific configuration could be provided by a workspace template for exemple.

 A templating plugin can be defined with a name, a set of template paths and a set of data-model paths (which can be a json file, a property file, ... any file that defines variables). We can then trigger the generation of file by specifying the template plugin name and an output path with a new instruction or an existing one (such as -includeresource).

A workspace template can then configure the template plugin to generate readme file based on the json metadata files (see #2139).

FreeMarker or any other template engine could be used, but I don't know if you allow 3rd party dependencies ? (FreeMarker has no dependency)

**Comment by cdecode** [comment]
See issue #2138 and PR #2169
